File: web/js/foundQuote.js
// foundQuote.js -- a quiet aside every few minutes of active reading: a real
// sentence pulled from an article the reader has actually opened, never a
// canned quote and never AI-generated. It's a plain editorial device (the
// pull-quote a real newspaper sets in larger type to break up a long piece),
// not a reward for reading -- no counter, no streak, nothing to "unlock".
//
// Two presentations, chosen in Settings (house: "found quote" default):
//   - quote:    a small blockquote-style break, hairline rules above/below.
//   - dateline: a single wire-style line, SOURCE NAME in small caps then the
//               line, like a wire-service insert rather than a set-off block.
// Both read from the exact same picked sentence -- only the wrapping differs.

import { sanitizeHtml } from './sanitize.js';

const MIN_LEN = 50;
const MAX_LEN = 220;

// Good enough for picking a plausible pull-quote, not a linguistic parser:
// split on sentence-ending punctuation followed by whitespace and a capital
// letter (or an opening quote), then filter to a length that reads well set
// apart on its own.
function sentencesFrom(html) {
  const frag = sanitizeHtml(html, { allowImages: false });
  // frag.textContent alone glues adjacent block elements together with no
  // separator at all ("...come.Supporters say...") since <p> boundaries carry
  // no whitespace of their own -- joining each top-level node's own text with
  // a space keeps paragraph breaks from silently vanishing into one run-on
  // blob the length filter below would then reject outright.
  const text = Array.from(frag.childNodes)
    .map((node) => (node.textContent || '').trim())
    .filter(Boolean)
    .join(' ')
    .replace(/\s+/g, ' ')
    .trim();
  if (!text) return [];
  return text
    .split(/(?<=[.!?])\s+(?=[A-Z"“])/)
    .map((s) => s.trim())
    .filter((s) => s.length >= MIN_LEN && s.length <= MAX_LEN)
    .filter((s) => !/https?:\/\//.test(s))
    .filter((s) => (s.match(/[A-Z]{5,}/g) || []).length === 0); // skip all-caps runs (credits/bylines that slipped through)
}

/**
 * Picks one real sentence from something the reader has actually read.
 * Prefers whichever item is currently open (the quote then reads as "pulled
 * from the piece in front of you"); otherwise a random already-read item with
 * real extracted text. Returns null rather than reaching for thin content --
 * no candidate this cycle just means no aside this cycle.
 */
export function pickFoundQuote(state) {
  const items = state.items || [];
  const readIds = state.readIds || new Set();
  const articles = state.articles || {};

  const candidates = items.filter((it) => {
    if (!readIds.has(it.id)) return false;
    const extracted = articles[it.id];
    return Boolean((extracted && extracted.html) || it.contentHtml);
  });
  if (!candidates.length) return null;

  const current = candidates.find((it) => it.id === state.currentItemId);
  const pool = current ? [current] : candidates;
  const item = pool[Math.floor(Math.random() * pool.length)];

  const extracted = articles[item.id];
  const html = (extracted && extracted.html) || item.contentHtml;
  const sentences = sentencesFrom(html);
  if (!sentences.length) return null;

  const source = (state.sources || []).find((s) => s.id === item.sourceId);
  return {
    itemId: item.id,
    text: sentences[Math.floor(Math.random() * sentences.length)],
    sourceTitle: (source && source.title) || item.author || 'Unknown source',
  };
}

export function buildFoundQuoteAside(quote, style) {
  const aside = document.createElement('aside');
  aside.setAttribute('aria-label', 'A line from what you have been reading');

  if (style === 'dateline') {
    aside.className = 'nooz-aside-dateline';
    const p = document.createElement('p');
    const tag = document.createElement('span');
    tag.className = 'nooz-aside-dateline-tag';
    tag.textContent = quote.sourceTitle.toUpperCase();
    p.appendChild(tag);
    const rest = document.createElement('span');
    rest.textContent = ` — “${quote.text}”`;
    p.appendChild(rest);
    aside.appendChild(p);
    return aside;
  }

  aside.className = 'nooz-aside-quote';
  const rule1 = document.createElement('hr');
  rule1.className = 'nooz-divider nooz-aside-quote-rule';
  aside.appendChild(rule1);
  const q = document.createElement('p');
  q.className = 'nooz-aside-quote-text';
  q.textContent = `“${quote.text}”`;
  aside.appendChild(q);
  const attr = document.createElement('p');
  attr.className = 'nooz-aside-quote-attr';
  attr.textContent = quote.sourceTitle;
  aside.appendChild(attr);
  const rule2 = document.createElement('hr');
  rule2.className = 'nooz-divider nooz-aside-quote-rule';
  aside.appendChild(rule2);
  return aside;
}

/** Splices the aside in after the paragraph roughly at the middle of the body -- a stable, deterministic spot, the way a real pull-quote breaks up a long column. */
export function insertFoundQuoteAside(bodyEl, asideNode) {
  const paras = Array.from(bodyEl.children).filter((el) => el.tagName === 'P');
  if (!paras.length) {
    bodyEl.appendChild(asideNode);
    return;
  }
  paras[Math.floor(paras.length / 2)].after(asideNode);
}

File: feature/reader/src/main/kotlin/xyz/mdhv/riverwip/feature/reader/FoundQuote.kt
package xyz.mdhv.riverwip.feature.reader

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import xyz.mdhv.riverwip.design.R as DesignR
import xyz.mdhv.riverwip.design.Tokens
import xyz.mdhv.riverwip.model.ReadingAsideStyle

/**
 * A quiet aside every so often while reading (owner's ask, matching the web
 * reader's identical feature): a real sentence pulled from the article
 * actually open, never fabricated, never a reward -- no counter, no streak,
 * just an editorial break the way a long newspaper column gets a pull-quote.
 * The reading clock and picker live in [ReaderViewModel]; this file only
 * holds the picking logic and the two presentations.
 */
data class FoundQuote(val itemId: String, val text: String, val sourceTitle: String)

private const val MIN_LEN = 50
private const val MAX_LEN = 220
private val CAPS_RUN = Regex("[A-Z]{5,}")
private val SENTENCE_SPLIT = Regex("(?<=[.!?])\\s+(?=[A-Z\"“])")

private fun sentencesFrom(paragraphs: List<String>): List<String> {
    val text = paragraphs.joinToString(" ") { it.trim() }.replace(Regex("\\s+"), " ").trim()
    if (text.isEmpty()) return emptyList()
    return text.split(SENTENCE_SPLIT)
        .map { it.trim() }
        .filter { it.length in MIN_LEN..MAX_LEN }
        .filter { !it.contains("http://") && !it.contains("https://") }
        .filter { !CAPS_RUN.containsMatchIn(it) } // skip all-caps runs (credits/bylines that slipped through)
}

/** Picks one real sentence from the article currently open. Null if nothing in it reads well set apart on its own -- no aside this cycle, rather than reaching for thin content. */
fun pickFoundQuote(itemId: String, sourceTitle: String?, paragraphs: List<String>): FoundQuote? {
    val sentences = sentencesFrom(paragraphs)
    if (sentences.isEmpty()) return null
    return FoundQuote(itemId, sentences.random(), sourceTitle ?: "Unknown source")
}

@Composable
fun FoundQuoteAside(quote: FoundQuote, style: ReadingAsideStyle, modifier: Modifier = Modifier) {
    if (style == ReadingAsideStyle.DATELINE) {
        Text(
            buildAnnotatedString {
                withStyle(SpanStyle(fontWeight = FontWeight.Bold)) { append(quote.sourceTitle.uppercase()) }
                append(" — “${quote.text}”")
            },
            style = MaterialTheme.typography.bodyMedium.copy(fontStyle = FontStyle.Italic),
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = modifier.fillMaxWidth(),
        )
        return
    }

    Column(modifier.fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) {
        HorizontalDivider(modifier = Modifier.width(48.dp), color = MaterialTheme.colorScheme.outlineVariant)
        Text(
            stringResource(DesignR.string.reader_quote, quote.text),
            style = MaterialTheme.typography.titleMedium.copy(fontStyle = FontStyle.Italic),
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(vertical = Tokens.Spacing.sm),
        )
        Text(
            quote.sourceTitle,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        HorizontalDivider(
            modifier = Modifier.width(48.dp).padding(top = Tokens.Spacing.sm),
            color = MaterialTheme.colorScheme.outlineVariant,
        )
    }
}
